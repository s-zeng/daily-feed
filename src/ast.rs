use nonempty::NonEmpty;
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct ArticleContent {
    pub blocks: NonEmpty<ContentBlock>,
    pub reading_time_minutes: u32,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct FeedContent {
    pub articles: NonEmpty<Article>,
    pub total_reading_time_minutes: u32,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct DocumentContent {
    pub feeds: NonEmpty<Feed>,
    pub total_reading_time_minutes: u32,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct Headline {
    pub title: String,
    pub published_date: Option<String>,
    pub source_name: String,
    pub url: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct Document {
    pub metadata: DocumentMetadata,
    pub front_page: Option<Vec<ContentBlock>>,
    pub content: Option<DocumentContent>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct DocumentMetadata {
    pub title: String,
    pub author: String,
    pub description: Option<String>,
    pub generated_at: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct Feed {
    pub name: String,
    pub description: Option<String>,
    pub url: Option<String>,
    pub content: Option<FeedContent>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct Article {
    pub title: String,
    pub metadata: ArticleMetadata,
    pub comments: Vec<Comment>,
    pub content: Option<ArticleContent>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct ArticleMetadata {
    pub published_date: Option<String>,
    pub author: Option<String>,
    pub url: Option<String>,
    pub feed_name: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct Comment {
    pub author: String,
    pub content: Vec<ContentBlock>,
    pub upvotes: u32,
    pub downvotes: u32,
    pub timestamp: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub enum ContentBlock {
    Paragraph(TextContent),
    Heading {
        level: u8,
        content: TextContent,
    },
    List {
        ordered: bool,
        items: Vec<TextContent>,
    },
    Quote(TextContent),
    Code {
        language: Option<String>,
        content: String,
    },
    Link {
        url: String,
        text: String,
    },
    Image {
        url: String,
        alt: Option<String>,
    },
    Raw(String), // For complex HTML that we want to preserve as-is
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct TextContent {
    pub spans: Vec<TextSpan>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct TextSpan {
    pub text: String,
    pub formatting: TextFormatting,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Default)]
pub struct TextFormatting {
    pub bold: bool,
    pub italic: bool,
    pub code: bool,
    pub link: Option<String>,
}

impl Document {
    pub fn new(title: String, author: String) -> Self {
        Self {
            metadata: DocumentMetadata {
                title,
                author,
                description: None,
                generated_at: chrono::Utc::now().to_rfc3339(),
            },
            front_page: None,
            content: None,
        }
    }

    pub fn set_content(&mut self, feeds: NonEmpty<Feed>, total_reading_time_minutes: u32) {
        self.content = Some(DocumentContent {
            feeds,
            total_reading_time_minutes,
        });
    }

    pub fn set_front_page(&mut self, front_page: Vec<ContentBlock>) {
        self.front_page = Some(front_page);
    }

    pub fn total_articles(&self) -> usize {
        self.content
            .as_ref()
            .map(|c| c.feeds.iter().map(|f| f.content.as_ref().map_or(0, |fc| fc.articles.len())).sum())
            .unwrap_or(0)
    }

    pub fn extract_headlines(&self) -> Vec<Headline> {
        self.content
            .as_ref()
            .map(|c| {
                c.feeds
                    .iter()
                    .flat_map(|feed| {
                        feed.content.as_ref().map(|fc| {
                            fc.articles.iter().map(|article| Headline {
                                title: article.title.clone(),
                                published_date: article.metadata.published_date.clone(),
                                source_name: article.metadata.feed_name.clone(),
                                url: article.metadata.url.clone(),
                            }).collect::<Vec<_>>()
                        }).unwrap_or_default()
                    })
                    .collect()
            })
            .unwrap_or_default()
    }
}

impl Feed {
    pub fn new(name: String) -> Self {
        Self {
            name,
            description: None,
            url: None,
            content: None,
        }
    }

    pub fn with_description(mut self, description: String) -> Self {
        self.description = Some(description);
        self
    }

    pub fn with_url(mut self, url: String) -> Self {
        self.url = Some(url);
        self
    }

    pub fn set_content(&mut self, articles: NonEmpty<Article>, total_reading_time_minutes: u32) {
        self.content = Some(FeedContent {
            articles,
            total_reading_time_minutes,
        });
    }
}

impl Article {
    pub fn new(title: String, feed_name: String) -> Self {
        Self {
            title,
            metadata: ArticleMetadata {
                published_date: None,
                author: None,
                url: None,
                feed_name,
            },
            comments: Vec::new(),
            content: None,
        }
    }

    pub fn set_content(&mut self, blocks: NonEmpty<ContentBlock>, reading_time_minutes: u32) {
        self.content = Some(ArticleContent {
            blocks,
            reading_time_minutes,
        });
    }

    pub fn with_published_date(mut self, date: String) -> Self {
        self.metadata.published_date = Some(date);
        self
    }

    pub fn with_url(mut self, url: String) -> Self {
        self.metadata.url = Some(url);
        self
    }

    pub fn add_comment(&mut self, comment: Comment) {
        self.comments.push(comment);
    }
}

impl TextContent {
    pub fn plain(text: String) -> Self {
        Self {
            spans: vec![TextSpan {
                text,
                formatting: TextFormatting::default(),
            }],
        }
    }

    pub fn from_spans(spans: Vec<TextSpan>) -> Self {
        Self { spans }
    }

    pub fn is_empty(&self) -> bool {
        self.spans.iter().all(|span| span.text.trim().is_empty())
    }

    pub fn to_plain_text(&self) -> String {
        self.spans.iter().map(|span| span.text.as_str()).collect()
    }
}

impl TextSpan {
    pub fn plain(text: String) -> Self {
        Self {
            text,
            formatting: TextFormatting::default(),
        }
    }

    pub fn bold(text: String) -> Self {
        Self {
            text,
            formatting: TextFormatting {
                bold: true,
                ..Default::default()
            },
        }
    }

    pub fn italic(text: String) -> Self {
        Self {
            text,
            formatting: TextFormatting {
                italic: true,
                ..Default::default()
            },
        }
    }

    pub fn code(text: String) -> Self {
        Self {
            text,
            formatting: TextFormatting {
                code: true,
                ..Default::default()
            },
        }
    }

    pub fn link(text: String, url: String) -> Self {
        Self {
            text,
            formatting: TextFormatting {
                link: Some(url),
                ..Default::default()
            },
        }
    }
}
