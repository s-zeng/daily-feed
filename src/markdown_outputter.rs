use crate::ast::*;
use std::error::Error;
use std::fs;
use std::path::Path;

pub struct MarkdownOutputter;

impl MarkdownOutputter {
    pub fn new() -> Self {
        Self
    }

    pub fn generate_markdown(&self, document: &Document, output_filename: &str) -> Result<(), Box<dyn Error>> {
        let markdown_content = self.render_document_to_markdown(document)?;
        
        // Ensure the output directory exists
        if let Some(parent) = Path::new(output_filename).parent() {
            fs::create_dir_all(parent)?;
        }
        
        fs::write(output_filename, markdown_content)?;
        Ok(())
    }

    fn render_document_to_markdown(&self, document: &Document) -> Result<String, Box<dyn Error>> {
        let mut markdown = String::new();
        
        // Document header
        markdown.push_str(&format!("# {}\n\n", document.metadata.title));
        
        if let Some(description) = &document.metadata.description {
            markdown.push_str(&format!("{}\n\n", description));
        }
        
        markdown.push_str(&format!("**Author:** {}\n", document.metadata.author));
        markdown.push_str(&format!("**Generated:** {}\n", document.metadata.generated_at));
        markdown.push_str(&format!("**Total Articles:** {}\n\n", document.total_articles()));
        
        // Table of contents
        markdown.push_str("## Table of Contents\n\n");
        for feed in &document.feeds {
            markdown.push_str(&format!("- [{}](#{})\n", feed.name, self.to_anchor(&feed.name)));
            for article in &feed.articles {
                markdown.push_str(&format!("  - [{}](#{})\n", 
                    article.title, 
                    self.to_anchor(&article.title)
                ));
            }
        }
        markdown.push_str("\n---\n\n");
        
        // Feed sections
        for feed in &document.feeds {
            markdown.push_str(&self.render_feed_to_markdown(feed)?);
        }
        
        Ok(markdown)
    }

    fn render_feed_to_markdown(&self, feed: &Feed) -> Result<String, Box<dyn Error>> {
        let mut markdown = String::new();
        
        markdown.push_str(&format!("## {}\n\n", feed.name));
        
        if let Some(description) = &feed.description {
            markdown.push_str(&format!("{}\n\n", description));
        }
        
        markdown.push_str(&format!("**Total Articles:** {}\n\n", feed.articles.len()));
        
        for article in &feed.articles {
            markdown.push_str(&self.render_article_to_markdown(article)?);
            markdown.push_str("\n---\n\n");
        }
        
        Ok(markdown)
    }

    fn render_article_to_markdown(&self, article: &Article) -> Result<String, Box<dyn Error>> {
        let mut markdown = String::new();
        
        // Article header
        markdown.push_str(&format!("### {}\n\n", article.title));
        
        // Metadata
        if let Some(date) = &article.metadata.published_date {
            markdown.push_str(&format!("**Published:** {}\n", date));
        }
        if let Some(author) = &article.metadata.author {
            markdown.push_str(&format!("**Author:** {}\n", author));
        }
        markdown.push_str(&format!("**Source:** {}\n", article.metadata.feed_name));
        if let Some(url) = &article.metadata.url {
            markdown.push_str(&format!("**Link:** [Read original article]({})\n", url));
        }
        markdown.push_str("\n");
        
        // Content
        for block in &article.content {
            markdown.push_str(&self.render_content_block_to_markdown(block)?);
        }
        
        // Comments
        if !article.comments.is_empty() {
            markdown.push_str("\n#### Top Comments\n\n");
            for comment in &article.comments {
                markdown.push_str(&self.render_comment_to_markdown(comment)?);
            }
        }
        
        Ok(markdown)
    }

    fn render_comment_to_markdown(&self, comment: &Comment) -> Result<String, Box<dyn Error>> {
        let mut markdown = String::new();
        
        markdown.push_str(&format!("> **{}** (Score: {})\n", comment.author, comment.score));
        if let Some(timestamp) = &comment.timestamp {
            markdown.push_str(&format!("> *{}*\n", timestamp));
        }
        markdown.push_str(">\n");
        
        for block in &comment.content {
            let block_markdown = self.render_content_block_to_markdown(block)?;
            // Prefix each line with "> " for blockquote formatting
            for line in block_markdown.lines() {
                if line.trim().is_empty() {
                    markdown.push_str(">\n");
                } else {
                    markdown.push_str(&format!("> {}\n", line));
                }
            }
        }
        markdown.push_str("\n");
        
        Ok(markdown)
    }

    fn render_content_block_to_markdown(&self, block: &ContentBlock) -> Result<String, Box<dyn Error>> {
        match block {
            ContentBlock::Paragraph(content) => {
                Ok(format!("{}\n\n", self.render_text_content_to_markdown(content)?))
            }
            ContentBlock::Heading { level, content } => {
                let heading_prefix = "#".repeat(*level as usize + 3); // +3 because document starts at h1, feeds at h2, articles at h3
                Ok(format!("{} {}\n\n", heading_prefix, self.render_text_content_to_markdown(content)?))
            }
            ContentBlock::List { ordered, items } => {
                let mut list_markdown = String::new();
                for (i, item) in items.iter().enumerate() {
                    let prefix = if *ordered {
                        format!("{}. ", i + 1)
                    } else {
                        "- ".to_string()
                    };
                    list_markdown.push_str(&format!("{}{}\n", prefix, self.render_text_content_to_markdown(item)?));
                }
                list_markdown.push('\n');
                Ok(list_markdown)
            }
            ContentBlock::Quote(content) => {
                let quote_text = self.render_text_content_to_markdown(content)?;
                let mut quoted = String::new();
                for line in quote_text.lines() {
                    quoted.push_str(&format!("> {}\n", line));
                }
                quoted.push('\n');
                Ok(quoted)
            }
            ContentBlock::Code { language, content } => {
                let lang = language.as_deref().unwrap_or("");
                Ok(format!("```{}\n{}\n```\n\n", lang, content))
            }
            ContentBlock::Link { url, text } => {
                Ok(format!("[{}]({})\n\n", text, url))
            }
            ContentBlock::Image { url, alt } => {
                let alt_text = alt.as_deref().unwrap_or("");
                Ok(format!("![{}]({})\n\n", alt_text, url))
            }
            ContentBlock::Raw(html) => {
                // For raw HTML, we could try to convert it, but for now just wrap it
                Ok(format!("```html\n{}\n```\n\n", html))
            }
        }
    }

    fn render_text_content_to_markdown(&self, content: &TextContent) -> Result<String, Box<dyn Error>> {
        let mut markdown = String::new();
        
        for span in &content.spans {
            let mut text = span.text.clone();
            
            // Apply formatting
            if span.formatting.code {
                text = format!("`{}`", text);
            }
            if span.formatting.bold {
                text = format!("**{}**", text);
            }
            if span.formatting.italic {
                text = format!("*{}*", text);
            }
            if let Some(url) = &span.formatting.link {
                text = format!("[{}]({})", text, url);
            }
            
            markdown.push_str(&text);
        }
        
        Ok(markdown)
    }

    fn to_anchor(&self, text: &str) -> String {
        text.to_lowercase()
            .replace(' ', "-")
            .replace(['(', ')', '[', ']', '{', '}', '<', '>', '"', '\'', '/', '\\', '|', '?', '*', '&', '%', '$', '#', '@', '!', '^', '~', '`', '+', '=', ',', '.', ';', ':'], "")
            .chars()
            .filter(|c| c.is_alphanumeric() || *c == '-' || *c == '_')
            .collect()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_render_text_content() {
        let outputter = MarkdownOutputter::new();
        
        let content = TextContent::from_spans(vec![
            TextSpan::plain("Hello ".to_string()),
            TextSpan::bold("world".to_string()),
            TextSpan::plain("!".to_string()),
        ]);
        
        let markdown = outputter.render_text_content_to_markdown(&content).unwrap();
        assert_eq!(markdown, "Hello **world**!");
    }

    #[test]
    fn test_render_paragraph() {
        let outputter = MarkdownOutputter::new();
        
        let block = ContentBlock::Paragraph(TextContent::plain("Test paragraph".to_string()));
        let markdown = outputter.render_content_block_to_markdown(&block).unwrap();
        assert_eq!(markdown, "Test paragraph\n\n");
    }

    #[test]
    fn test_render_heading() {
        let outputter = MarkdownOutputter::new();
        
        let block = ContentBlock::Heading {
            level: 1,
            content: TextContent::plain("Test Heading".to_string()),
        };
        let markdown = outputter.render_content_block_to_markdown(&block).unwrap();
        assert_eq!(markdown, "#### Test Heading\n\n");
    }

    #[test]
    fn test_to_anchor() {
        let outputter = MarkdownOutputter::new();
        
        assert_eq!(outputter.to_anchor("Hello World"), "hello-world");
        assert_eq!(outputter.to_anchor("Test & More"), "test--more");
        assert_eq!(outputter.to_anchor("Complex (Test) [Case]!"), "complex-test-case");
    }

    #[test]
    fn test_render_code_block() {
        let outputter = MarkdownOutputter::new();
        
        let block = ContentBlock::Code {
            language: Some("rust".to_string()),
            content: "fn main() {\n    println!(\"Hello\");\n}".to_string(),
        };
        let markdown = outputter.render_content_block_to_markdown(&block).unwrap();
        assert_eq!(markdown, "```rust\nfn main() {\n    println!(\"Hello\");\n}\n```\n\n");
    }

    #[test]
    fn test_render_list() {
        let outputter = MarkdownOutputter::new();
        
        let block = ContentBlock::List {
            ordered: false,
            items: vec![
                TextContent::plain("Item 1".to_string()),
                TextContent::plain("Item 2".to_string()),
            ],
        };
        let markdown = outputter.render_content_block_to_markdown(&block).unwrap();
        assert_eq!(markdown, "- Item 1\n- Item 2\n\n");
    }
}