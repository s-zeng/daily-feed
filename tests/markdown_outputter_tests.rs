use daily_feed::markdown_outputter::*;
use daily_feed::ast::*;

#[test]
fn test_render_text_content() {
    let outputter = MarkdownOutputter::new();
    
    let content = TextContent::from_spans(vec![
        TextSpan::plain("Hello ".to_string()),
        TextSpan::bold("world".to_string()),
        TextSpan::plain("!".to_string()),
    ]);
    
    let markdown = outputter.render_text_content_to_markdown(&content).unwrap();
    insta::assert_snapshot!(markdown);
}

#[test]
fn test_render_paragraph() {
    let outputter = MarkdownOutputter::new();
    
    let block = ContentBlock::Paragraph(TextContent::plain("Test paragraph".to_string()));
    let markdown = outputter.render_content_block_to_markdown(&block).unwrap();
    insta::assert_snapshot!(markdown);
}

#[test]
fn test_render_heading() {
    let outputter = MarkdownOutputter::new();
    
    let block = ContentBlock::Heading {
        level: 1,
        content: TextContent::plain("Test Heading".to_string()),
    };
    let markdown = outputter.render_content_block_to_markdown(&block).unwrap();
    insta::assert_snapshot!(markdown);
}

#[test]
fn test_to_anchor() {
    let outputter = MarkdownOutputter::new();
    
    let result1 = outputter.to_anchor("Hello World");
    insta::assert_snapshot!(result1, @"hello-world");
}

#[test]
fn test_to_anchor_special_chars() {
    let outputter = MarkdownOutputter::new();
    
    let result = outputter.to_anchor("Test & More");
    insta::assert_snapshot!(result, @"test--more");
}

#[test]
fn test_to_anchor_complex() {
    let outputter = MarkdownOutputter::new();
    
    let result = outputter.to_anchor("Complex (Test) [Case]!");
    insta::assert_snapshot!(result, @"complex-test-case");
}

#[test]
fn test_render_code_block() {
    let outputter = MarkdownOutputter::new();
    
    let block = ContentBlock::Code {
        language: Some("rust".to_string()),
        content: "fn main() {\n    println!(\"Hello\");\n}".to_string(),
    };
    let markdown = outputter.render_content_block_to_markdown(&block).unwrap();
    insta::assert_snapshot!(markdown);
}

#[test]
fn test_render_list() {
    let outputter = MarkdownOutputter::new();
    
    let block = ContentBlock::List {
        ordered: false,
        items: vec![
            TextContent::plain("Item 1".to_string()),
            TextContent::plain("Item 2".to_string()),
        ],
    };
    let markdown = outputter.render_content_block_to_markdown(&block).unwrap();
    insta::assert_snapshot!(markdown);
}

#[test]
fn test_render_document_with_front_page() {
    let outputter = MarkdownOutputter::new();
    
    let article = Article {
        title: "Test Article".to_string(),
        content: vec![ContentBlock::Paragraph(TextContent::plain("Article content".to_string()))],
        metadata: ArticleMetadata {
            published_date: Some("2025-01-01T00:00:00.000000Z".to_string()),
            author: Some("Test Author".to_string()),
            url: Some("https://example.com/article".to_string()),
            feed_name: "Test Feed".to_string(),
        },
        comments: vec![],
    };
    
    let feed = Feed {
        name: "Test Feed".to_string(),
        description: Some("A test feed".to_string()),
        url: Some("https://example.com/feed".to_string()),
        articles: vec![article],
    };
    
    let document = Document {
        metadata: DocumentMetadata {
            title: "Test Document".to_string(),
            author: "Test Author".to_string(),
            description: Some("Test description".to_string()),
            generated_at: "2025-01-01T00:00:00.000000Z".to_string(),
        },
        front_page: Some("This is a front page summary with important news highlights.".to_string()),
        feeds: vec![feed],
    };
    
    let markdown = outputter.render_document_to_markdown(&document).unwrap();
    insta::assert_snapshot!(markdown);
}

#[test]
fn test_render_document_without_front_page() {
    let outputter = MarkdownOutputter::new();
    
    let article = Article {
        title: "Test Article".to_string(),
        content: vec![ContentBlock::Paragraph(TextContent::plain("Article content".to_string()))],
        metadata: ArticleMetadata {
            published_date: Some("2025-01-01T00:00:00.000000Z".to_string()),
            author: Some("Test Author".to_string()),
            url: Some("https://example.com/article".to_string()),
            feed_name: "Test Feed".to_string(),
        },
        comments: vec![],
    };
    
    let feed = Feed {
        name: "Test Feed".to_string(),
        description: Some("A test feed".to_string()),
        url: Some("https://example.com/feed".to_string()),
        articles: vec![article],
    };
    
    let document = Document {
        metadata: DocumentMetadata {
            title: "Test Document".to_string(),
            author: "Test Author".to_string(),
            description: Some("Test description".to_string()),
            generated_at: "2025-01-01T00:00:00.000000Z".to_string(),
        },
        front_page: None,
        feeds: vec![feed],
    };
    
    let markdown = outputter.render_document_to_markdown(&document).unwrap();
    insta::assert_snapshot!(markdown);
}

#[test]
fn test_front_page_multiline_content() {
    let outputter = MarkdownOutputter::new();
    
    let article = Article {
        title: "Test Article".to_string(),
        content: vec![ContentBlock::Paragraph(TextContent::plain("Article content".to_string()))],
        metadata: ArticleMetadata {
            published_date: Some("2025-01-01T00:00:00.000000Z".to_string()),
            author: Some("Test Author".to_string()),
            url: Some("https://example.com/article".to_string()),
            feed_name: "Test Feed".to_string(),
        },
        comments: vec![],
    };
    
    let feed = Feed {
        name: "Test Feed".to_string(),
        description: Some("A test feed".to_string()),
        url: Some("https://example.com/feed".to_string()),
        articles: vec![article],
    };
    
    let document = Document {
        metadata: DocumentMetadata {
            title: "Test Document".to_string(),
            author: "Test Author".to_string(),
            description: Some("Test description".to_string()),
            generated_at: "2025-01-01T00:00:00.000000Z".to_string(),
        },
        front_page: Some("# Breaking News\n\nMultiple important stories today:\n\n- Economic markets show volatility\n- Technology sector announces breakthrough\n- Climate summit reaches agreement".to_string()),
        feeds: vec![feed],
    };
    
    let markdown = outputter.render_document_to_markdown(&document).unwrap();
    insta::assert_snapshot!(markdown);
}