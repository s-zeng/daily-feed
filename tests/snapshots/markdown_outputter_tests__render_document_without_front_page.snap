---
source: tests/markdown_outputter_tests.rs
expression: markdown
---
# Test Document

Test description

**Author:** Test Author
**Generated:** 2025-01-01T00:00:00.000000Z
**Total Articles:** 1

## Table of Contents

- [Test Feed](#test-feed)
  - [Test Article](#test-article)

---

## Test Feed

A test feed

**Total Articles:** 1

### Test Article

**Published:** 2025-01-01T00:00:00.000000Z
**Author:** Test Author
**Source:** Test Feed
**Link:** [Read original article](https://example.com/article)

Article content


---
